### PR TITLE
Re-enable support for using the system libunwind

### LINKED ISSUE #42661: singlefilehost doesn't link to system libunwind
Trying to build runtime (commit 20710bbcae006e32f8a133c372c8d78722890982) with `-DCLR_CMAKE_USE_SYSTEM_LIBUNWIND` is currently broken:

```
$ ./build.sh -cmakeargs -DCLR_CMAKE_USE_SYSTEM_LIBUNWIND=TRUE 
...
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): in function `UnwindContextToWinContext(unw_cursor*, _CONTEXT*)':                                                      
  /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:176: undefined reference to `_ULx86_64_get_reg'                                                                                                          
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:177: undefined reference to `_ULx86_64_get_reg'                                                                                             
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:178: undefined reference to `_ULx86_64_get_reg'                                                                                             
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:179: undefined reference to `_ULx86_64_get_reg'                                                                                             
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:180: undefined reference to `_ULx86_64_get_reg'                                                                                             
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o):/home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:181: more undefined references to `_ULx86_64_get_reg' follow                                                                                                                                                                                                                      
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): in function `GetContextPointer(unw_cursor*, ucontext_t*, int, unsigned long**)':                                      
  /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:227: undefined reference to `_ULx86_64_get_save_loc'                                                                                                     
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): in function `PAL_VirtualUnwind':                                                                                      
  /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:340: undefined reference to `_ULx86_64_init_local'                                                                                                       
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:351: undefined reference to `_ULx86_64_step'                                                                                                
  /usr/bin/ld: /home/omajid/devel/dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:360: undefined reference to `_ULx86_64_is_signal_frame'                                                                                     
  clang-10: error: linker command failed with exit code 1 (use -v to see invocation)                                                                                                                                                          
  gmake[2]: *** [cli/apphost/static/CMakeFiles/singlefilehost.dir/build.make:282: cli/apphost/static/singlefilehost] Error 1                                                                                                                  
  gmake[1]: *** [CMakeFiles/Makefile2:510: cli/apphost/static/CMakeFiles/singlefilehost.dir/all] Error 2                                 
```

### Configuration

* Which version of .NET is the code running on?

dotnet/runtime commit `20710bbcae006e32f8a133c372c8d78722890982` as well as `release/5.0-preview8` tag.

* What OS and version, and what distro if applicable?

Fedora 32.

* What is the architecture (x64, x86, ARM, ARM64)?

x64, but probably affects everything.

* Do you know whether it is specific to that configuration?

Yes: building without `-DCLR_CMAKE_USE_SYSTEM_LIBUNWIND` works

### Regression?

Yes. This build configuration was supported in .NET Core 3.1 and we were using it in source-build to build on Fedora/RHEL. See https://github.com/dotnet/coreclr/pull/17164 and https://github.com/dotnet/source-build/pull/391

### Other information

The fix is probably the along the lines of https://github.com/dotnet/runtime/pull/42094: the singlefilehost needs to link to the system libunwind library.

### PR DESCRIPTION
**Pull request by omajid**
You can now build runtime against the system libunwind using:

    ./build.sh -cmakeargs -DCLR_CMAKE_USE_SYSTEM_LIBUNWIND

This allows Linux distributions that already ship a compatible version of libunwind library to use that instead of carrying a duplicate in .NET. This provides some benefits to them, including smaller build sizes, slightly faster builds and fewer places to fix any vulnerabilities

This functionality was already supported in .NET Core 3.1 and has regressed since.

CoreCLR already handles `-DCLR_CMAKE_USE_SYSTEM_LIBUNWIND`, so no changes are needed there.

The libraries build doesn't care about this cmake variable, but cmake itself fails if the variable is not used:

    EXEC : CMake error : [runtime/src/libraries/Native/build-native.proj]
        Manually-specified variables were not used by the project:

          CLR_CMAKE_USE_SYSTEM_LIBUNWIND

So libraries just needs to check and ignore this variable.

The singlefilehost needs to link against libunwind too. Otherwise the linker fails to resolve symbols, making the build fail:

    /usr/bin/ld: runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): in function `UnwindContextToWinContext(unw_cursor*, _CONTEXT*)':
      dotnet/runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:176: undefined reference to `_ULx86_64_get_reg'
    /usr/bin/ld: runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:177: undefined reference to `_ULx86_64_get_reg'
    /usr/bin/ld: runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:178: undefined reference to `_ULx86_64_get_reg'
    /usr/bin/ld: runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:179: undefined reference to `_ULx86_64_get_reg'
    /usr/bin/ld: runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:180: undefined reference to `_ULx86_64_get_reg'
    /usr/bin/ld: runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:181: more undefined references to `_ULx86_64_get_reg' follow
    /usr/bin/ld: runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): in function `GetContextPointer(unw_cursor*, ucontext_t*, int, unsigned long**)':
      runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:227: undefined reference to `_ULx86_64_get_save_loc'
    /usr/bin/ld: runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib/libcoreclrpal.a(seh.cpp.o): in function `PAL_VirtualUnwind':
      runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:340: undefined reference to `_ULx86_64_init_local'
    /usr/bin/ld: runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:351: undefined reference to `_ULx86_64_step'
    /usr/bin/ld: runtime/src/coreclr/src/pal/src/exception/seh-unwind.cpp:360: undefined reference to `_ULx86_64_is_signal_frame'
    clang-10: error: linker command failed with exit code 1 (use -v to see invocation)

Fixes: #42661

### CODE REVIEW

**Comment by omajid** [comment]
cc @janvorli

**Comment by omajid** [comment]
I am still trying to test this end-to-end.

File: src/libraries/Native/Unix/configure.cmake
Comment (omajid) [inline src/libraries/Native/Unix/configure.cmake:45]: source-build builds each component (coreclr, libraries, installer) separately and passes a separate set of args. It can skip passing `-DCLR_CMAKE_USE_SYSTEM_LIBUNWIND` to libraries. But this seems like a nice to have to make it possible to build all of runtime at once using the main `./build.sh` script in one call.

Comment (janvorli) [reply src/libraries/Native/Unix/configure.cmake:45]: It is kind of unfortunate that the main build script only supports single set of cmake args that it pushes to all three sub-parts (coreclr, installer, libraries). It would be nice to enhance it to enable specifying these flags separately for coreclr, installer and libraries so that we don't have to do tricks like this. But for this PR, it is fine as is.

File: src/installer/corehost/cli/apphost/static/CMakeLists.txt
Comment (janvorli) [inline src/installer/corehost/cli/apphost/static/CMakeLists.txt:110]: A nit - can you please delete the now empty if / endif and move the comment above to the location below where you have placed the libunwind finding?

**Review by janvorli — APPROVED** [review]
LGTM modulo the nit

File: src/libraries/Native/Unix/configure.cmake
Comment (omajid) [reply src/libraries/Native/Unix/configure.cmake:45]: I filed https://github.com/dotnet/runtime/issues/42711 for tracking that

**Review by janvorli — APPROVED** [review]

**Comment by omajid** [comment]
Thanks for reviewing and merging this!

**Comment by jeffschwMSFT** [comment]
@omajid is this needed for .NET 5 GA?

**Comment by omajid** [comment]
We (Red Hat) would definitely like to make use of this feature in .NET 5 GA release. We are using it in 3.1 and this is a regression from .NET Core 3.1 to .NET 5.

Ideally we would add it to 5.0 GA (or post GA if GA is absolutely frozen for now).

If this doesn't get added to 5.0 GA, we still have some options: we can carry this as a patch in [source-build](https://github.com/dotnet/source-build) or in our RPM packages. But that gets reviewed less thoroughly and becomes harder to maintain as runtime gets additional fixes.

I think we should decide to handle this issue consistently with #42094 and #42415: either all go to 5.0 or none of them go to 5.0 GA.

**Comment by jeffschwMSFT** [comment]
@dleeapho thoughts?

**Pull request merged**
